File: src/ast/tree/function/function.ts
import type { Pass } from "../../../pass/pass";
import type { Block } from "../../node/block";
import type { Lambda } from "../../node/lambda";
import { Node } from "../../node/node";
import type { Type } from "../../node/type";
import type { FunctionDeclaration } from "../statement/functionDeclaration";
import type { Parameter } from "./parameter";

export class Function extends Node<FunctionDeclaration> {
    constructor(
        public name: string,
        public lambda: Lambda,
        public returnType: Type<any> | null,
        parent: FunctionDeclaration | null
    ) {
        super(parent);
    }

    public visit(pass: Pass): void {
        pass.visitFunction(this);
    }
}

File: src/main/kotlin/me/honkling/miniscript/stdlib/builder/Function.kt
package me.honkling.miniscript.stdlib.builder

import me.honkling.miniscript.parser.ast.Block
import me.honkling.miniscript.parser.ast.Type
import me.honkling.miniscript.parser.ast.function.Function
import me.honkling.miniscript.parser.ast.function.Parameter
import me.honkling.miniscript.parser.ast.stack.Environment
import me.honkling.miniscript.parser.ast.statement.NativeStatement

fun Environment.function(
    name: String,
    returnType: Type<*>? = null,
    vararg parameters: Pair<String, Type<*>>,
    executionBlock: NativeStatement.() -> Unit
): Function {
    val block = Block(miniScript, mutableListOf(), null, null)
    block.statements += NativeStatement(executionBlock, block)

    val parameters = parameters.map { Parameter(it.first, it.second, null, false, emptyLocation, null) }
    val function = Function(name, parameters.toMutableList(), returnType, block, false, emptyLocation, this)
    parameters.forEach { it.parent = function }
    block.parent = function

    return function
}